provider/cloudflare/cloudflare.go

```diff
@@ -364,23 +395,54 @@ func (p *CloudFlareProvider) submitChanges(ctx context.Context, changes []*cloud
 
 			resourceContainer := cloudflare.ZoneIdentifier(zoneID)
 			if change.Action == cloudFlareUpdate {
+				if recordTypeCustomHostnameSupported[change.ResourceRecord.Type] {
+					chID, oldCh := p.getCustomHostnameIDbyOrigin(chs, change.ResourceRecord.Name)
+					if chID == "" && change.CustomHostname.Hostname != "" {
+						log.WithFields(logFields).Infof("Adding custom hostname %v", change.CustomHostname.Hostname)
+						_, chErr := p.Client.CreateCustomHostname(ctx, zoneID, change.CustomHostname)
+						if chErr != nil {
+							failedChange = true
+							log.WithFields(logFields).Errorf("failed to add custom hostname %v: %v", change.CustomHostname.Hostname, chErr)
+						}
+					} else if chID != "" && oldCh != "" && change.CustomHostname.Hostname == "" {
+						log.WithFields(logFields).Infof("Removing custom hostname %v", change.CustomHostname.Hostname)
+						chErr := p.Client.DeleteCustomHostname(ctx, zoneID, chID)
+						if chErr != nil {
+							failedChange = true
+							log.WithFields(logFields).Errorf("failed to remove custom hostname %v: %v", change.CustomHostname.Hostname, chErr)
+						}
+					} else if chID != "" && change.CustomHostname.Hostname != "" && oldCh != change.CustomHostname.Hostname {
+						log.WithFields(logFields).Infof("Replacing custom hostname: %v/%v to %v", chID, oldCh, change.CustomHostname.Hostname)
+						chDelErr := p.Client.DeleteCustomHostname(ctx, zoneID, chID)
+						if chDelErr != nil {
+							failedChange = true
+							log.WithFields(logFields).Errorf("failed to remove custom hostname %v/%v: %v", chID, oldCh, chDelErr)
+						}
+						_, chAddErr := p.Client.CreateCustomHostname(ctx, zoneID, change.CustomHostname)
+						if chAddErr != nil {
+							failedChange = true
+							log.WithFields(logFields).Errorf("failed to add custom hostname %v: %v", change.CustomHostname.Hostname, chAddErr)
```

Comment:
Do you want to make the error statement something like `Failed to replace custom hostname` so that it is different and more specific to the error message on line 405

Reply:
All possible scenarios:
- adding record (DNS along with the custom hostname) - "failed to **create**"
- deleting record - "failed to **delete**"
- updating record:
  - custom hostname didn't not exist before - "failed to **add**"
  - custom hostname exists and needs to be removed - "failed to **remove**"
  - custom hostname needs to be changed (cloudflare doesn't allow to rename the existing hostname, recreate only):
    - "failed to **remove**"
    - "failed to **add**"

I couldn't come up with a clear message for the last 2, besides "failed to add/remove custom hostname **on replace**" or "failed to add custom hostname **on update**".
Do you think it would it be less confusing? I would be happy to rephrase it.

Reply:
Yeah, I think just adding the word `replacement` would differentiate it from line 405 and let them know that it was in the process of replacing the record and is probably in a bad state now since it deleted the record and couldn't recreate the record.

Reply:
updated, hope that helps, thanks!